### PR TITLE
Orbit and PanZoom camera controls

### PR DESCRIPTION
**Pull request by Korijn**
Ready to merge!

Changes:

* Orbit and panzoom camera controls
  * Right mouse button drag to pan
  * Left mouse button drag to rotate
  * Mouse wheel to zoom
  * Tuneable speeds for panning, zooming and rotating
  * Choose between zooming based on camera distance or zoom
  * Came up with an API that allows the user to remain in control over state flow and GUI glue
* Had to revise the orthographic camera control flow a little bit; previously, it would autonomously (i.e. without user input) zoom if you set ` zoom != 1`
* WorldObject.look_at utility
* Implement more functions on Spherical linalg class
* While debugging the above, I also:
  * Added an AxesHelper to debug coordinate systems
    * Since linesegments don't support vertex coloring yet, I had to create three separate objects to be able to set different colors on each axis
  * Fixed a bug in perspective and orthographic camera (top/bottom were flipped)
* Orbit camera example:

orbit

* Panzoom (zoom-to-mouse!) camera example:

panzoom

* Multi-slice example:

multislice

### CODE REVIEW

File: pygfx/linalg/spherical.py [hunk 1]
```diff
@@ -1,7 +1,36 @@
...
+            elif y > 1:
+                y = 1
+            self.phi = math.acos(y)
+        return self
+
+    def make_safe(self) -> "Spherical":
+        # restrict phi to (epsilon, pi-epsilon)
+        self.phi = max(self._eps, min(math.pi - self._eps, self.phi))
+        # restrict theta to (0, 2*pi)
+        self.theta = self.theta % self._2pi
```
Comment (Korijn) [inline pygfx/linalg/spherical.py:36]: Missing `return self`

File: pygfx/objects/_base.py [hunk 2]
```diff
@@ -77,3 +79,13 @@ def update_matrix_world(
...
+
+    def look_at(self, target: Vector3):
+        self.update_matrix_world(update_parents=True, update_children=False)
+        v = Vector3().set_from_matrix_position(self.matrix_world)
+        m = Matrix4().look_at(v, target, self.up)
+        self.rotation.set_from_rotation_matrix(m)
+        if self.parent:
+            m.extract_rotation(self.parent.matrix_world)
+            q = Quaternion().set_from_rotation_matrix(m)
+            self.rotation.premultiply(q.inverse())
```
Comment (Korijn) [inline pygfx/objects/_base.py:91]: Could reduce gc pressure with some temp vars preallocated

File: examples/orbit_camera.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+from PyQt5 import QtWidgets, QtCore
+from wgpu.gui.qt import WgpuCanvas
+
+
+class WgpuCanvasWithInputEvents(WgpuCanvas):
+    _flip_y = np.array([1, -1])
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.drag = {}
```
Comment (Korijn) [inline examples/orbit_camera.py:18]: What's the proper abstraction to decouple this?

File: examples/multi_slice.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+                [1.0, 0.0, z_idx],
+                [0.0, 1.0, z_idx],
+                [1.0, 1.0, z_idx],
+            ],
+            dtype=np.float32,
+        )
+    print(texcoords)
+    geometry.texcoords = gfx.Buffer(texcoords, usage="vertex|storage")
+
+    material = gfx.MeshVolumeSliceMaterial(map=view, clim=(0, 255))
```
Comment (Korijn) [inline examples/multi_slice.py:61]: Don't need to recreate this for every plane!

File: examples/panzoom_camera.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+        # apply zoom and compute change in dimensions of camera frustum
+        controls.zoom(event.angleDelta().y())
+        controls.update_camera(camera)
+        camera.update_bounds()
+        view = (camera.right - camera.left, camera.top - camera.bottom)
+        # compute new mouse position relative to widget center
+        relpos = tuple(fracpos[i] * view[i] for i in range(2))
+        # compute delta and pan accordingly
+        delta = tuple(relpos[i] - relpos_old[i] for i in range(2))
+        controls.pan(*delta)
```
Comment (Korijn) [inline examples/panzoom_camera.py:37]: would be nice to embed this logic in a controls class. should probably do that.

File: pygfx/cameras/_orthographic.py [hunk 6]
```diff
@@ -33,29 +31,35 @@ def __repr__(self) -> str:
...
+        self.height = float(height)
+        self._view_aspect = self.width / self.height
+        self.update_bounds()
 
-    def update_projection_matrix(self):
+    def update_bounds(self):
         # Get the reference width / height
+        # TODO: when can this lead to a change? unclear to me
         width = self.width / self.zoom
         height = self.height / self.zoom
```
Comment (Korijn) [inline pygfx/cameras/_orthographic.py:43]: if `self.zoom` is a single number, how could it ever result in a new aspect ratio?

File: pygfx/objects/_base.py [hunk 7]
```diff
@@ -77,3 +83,13 @@ def update_matrix_world(
         if update_children:
             for child in self._children:
                 child.update_matrix_world()
+
+    def look_at(self, target: Vector3):
```
Comment (almarklein) [inline pygfx/objects/_base.py:87]: Oh sweet, so we can orient any object this way 😮

File: pygfx/objects/_base.py [hunk 8]
```diff
@@ -12,6 +12,10 @@ class WorldObject:
     into a single empty world object.
     """
 
+    _v = Vector3()
+    _m = Matrix4()
+    _q = Quaternion()
```
[the remark this answers is not in the document]
Comment (Korijn) [reply pygfx/objects/_base.py:17]: Yes

File: pygfx/controls/_orbit.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
+from ..linalg import Vector3, Matrix4, Quaternion, Spherical
+
+
+class OrbitControls:
+    _m = Matrix4()
+    _v = Vector3()
+    _origin = Vector3()
+    _orbit_up = Vector3(0, 1, 0)
+    _s = Spherical()
```
[the remark this answers is not in the document]
Comment (Korijn) [reply pygfx/controls/_orbit.py:9]: Well. I don't know about that. Is it necessary to be thread safe?

File: pygfx/controls/_orbit.py [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+    def zoom(self, delta: float) -> "OrbitControls":
+        if self.zoom_type == "distance":
+            self.distance -= delta * self.zoom_speed
+            if self.distance < 0:
+                self.distance = 0
+        elif self.zoom_type == "zoom":
+            delta = delta * self.zoom_speed * 0.01
+            if self.zoom_ < 1.0:
+                delta *= self.zoom_
+            self.zoom_ += delta
```
[the remark this answers is not in the document]
Comment (Korijn) [reply pygfx/controls/_orbit.py:80]: Zooming out happens with zoom < 1.0, and if you keep incrementing by the same number of every tick of the scroll wheel, you'll make bigger and bigger jumps zooming out. It's very uncomfortable. Multiplying by `self.zoom_` makes the steps smaller and smaller to compensate.

File: pygfx/controls/_orbit.py [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+        return self
+
+    def pan(self, x: float, y: float) -> "OrbitControls":
+        self._v.set(x, -y, 0).multiply_scalar(self.pan_speed).apply_quaternion(
+            self.rotation
+        )
+        self.target.sub(self._v)
+        return self
+
+    def rotate(self, x: float, y: float) -> "OrbitControls":
```
[the remark this answers is not in the document]
Comment (Korijn) [reply pygfx/controls/_orbit.py:49]: no, they mean phi and theta :) haha. it depends on what convention you are used to

File: pygfx/controls/_orbit.py [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+        if eye is None:
+            eye = Vector3(50.0, 50.0, 50.0)
+        if target is None:
+            target = Vector3()
+        if up is None:
+            up = Vector3(0.0, 1.0, 0.0)
+        self.look_at(eye, target, up)
+        self.rotate_speed = 0.02
+        self.zoom_speed = 0.125
+        self.pan_speed = 1.0
```
[the remark this answers is not in the document]
Comment (Korijn) [reply pygfx/controls/_orbit.py:28]: alright, will remove them

File: examples/orbit_camera.py [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+    def mousePressEvent(self, event):  # noqa: N802
+        button = event.button()
+        mode = self._drag_modes.get(button, None)
+        if self.drag or not mode:
+            return  # drag is already initiated, or unknown button pressed
+        self.drag = {
+            "mode": mode,
+            "button": button,
+            "start": (event.x(), event.y()),
+        }
```
[the remark this answers is not in the document]
Comment (Korijn) [reply examples/orbit_camera.py:33]: black made me do it :)

File: pygfx/objects/_base.py (hunk 7 above)
Comment (Korijn) [reply pygfx/objects/_base.py:87]: yes! it also properly takes into account any parent transforms

File: examples/panzoom_camera.py [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+tex = gfx.Texture(im, dim=2, usage="sampled")
+geometry = gfx.PlaneGeometry(512, 512)
+material = gfx.MeshBasicMaterial(map=tex.get_view(filter="linear"), clim=(0, 255))
+plane = gfx.Mesh(geometry, material)
+plane.scale.y = -1
+scene.add(plane)
+
+camera = gfx.OrthographicCamera(512, 512)
+camera.position.set(0, 0, 500)
+controls = gfx.OrbitControls(camera.position.clone(), zoom="zoom")
```
[the remark this answers is not in the document]
Comment (Korijn) [reply examples/panzoom_camera.py:89]: yeah. I'll address a number of these comments simultaneously by introducing that

File: pygfx/cameras/_orthographic.py (hunk 6 above)
Comment (Korijn) [reply pygfx/cameras/_orthographic.py:43]: yes, but I'd say that we don't need to recompute the aspect ratio then, making the function a lot simpler

File: examples/multi_slice.py [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+material = gfx.MeshVolumeSliceMaterial(map=view, clim=(0, 255))
+
+for axis in [0, 1, 2]:
+    nslices = vol.shape[0]
+
+    # TODO: if we could set the slicing plane on the material parametrically
+    # we could reuse the same plane geometry for all slices here
+    # TODO: there seems to be a problem with the texcoords... only one plane looks correct
+    # TODO: why is the third axis not in [0..1] range?
+    # TODO: also add a mesh slice for each plane
```
[the remark this answers is not in the document]
Comment (almarklein) [reply examples/multi_slice.py]: Done. I don't think the material should have a slicing pane. I like how it simply uses the texcoords ....

File: pygfx/controls/_orbit.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (almarklein) [reply pygfx/controls/_orbit.py:80]: Shouldn't the zoom be multiplicative instead of additive all the time?

File: pygfx/controls/_orbit.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (almarklein) [reply pygfx/controls/_orbit.py:49]: I can live with `phi` and `theta`, but `x` and `y` seem a bit off ;)

File: examples/multi_slice.py [hunk 16]
```diff
@@ -6,74 +6,95 @@
...
     scene.add(plane)
 
 camera = gfx.OrthographicCamera(200, 200)
 camera.position.set(50, 50, 50)
 camera.look_at(gfx.linalg.Vector3())
+controls = gfx.OrbitControls(camera.position.clone(), up=gfx.linalg.Vector3(0, 0, 1))
 
 
 def animate():
+    # todo: should the control set it directly?
```
Comment (Korijn) [inline examples/multi_slice.py:96]: no, then the user would lose control over the control flow

File: pygfx/cameras/_orthographic.py (hunk 6 above)
Comment (almarklein) [reply pygfx/cameras/_orthographic.py:43]: I don't follow, you mean we don't need to call this method if only the zoom changes?

File: pygfx/controls/_orbit.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (almarklein) [reply pygfx/controls/_orbit.py:9]: Eventually someone will want to us threads ...

File: examples/multi_slice.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (Korijn) [reply examples/multi_slice.py]: Thanks for fixing this: I also updated the PR description

File: pygfx/controls/_orbit.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Korijn) [reply pygfx/controls/_orbit.py]: can you explain a little further what you mean?

File: pygfx/cameras/_orthographic.py (hunk 6 above)
Comment (Korijn) [reply pygfx/cameras/_orthographic.py:43]: we have this:

    def set_viewport_size(self, width, height):
        self.width = float(width)
        self.height = float(height)
        self._view_aspect = self.width / self.height

and this:

        width = self.width / self.zoom
        height = self.height / self.zoom
        # Increase either the width or height, depending on the view size
        aspect = width / height

and I don't see how that second `aspect` could ever be different from `self._view_aspect`

File: pygfx/controls/_orbit.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Korijn) [reply pygfx/controls/_orbit.py:9]: Well, I don't expect users to instantiate many controls objects, so I think making them instance local is fine here.

However we also have this approach in the linalg package. So there's that..

File: examples/multi_slice.py (hunk 16 above)
Comment (Korijn) [reply examples/multi_slice.py:100]: I kind of had to violate my own policy in the zoom_to_point code.. don't like it. :/ I'll see if I can implement it without having to mutate the camera

File: pygfx/controls/_orbit.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Korijn) [reply pygfx/controls/_orbit.py:9]: I've created #54 to track this

File: pygfx/controls/_orbit.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (almarklein) [reply pygfx/controls/_orbit.py]: I mean that perhaps we should always do `self.zoom *= delta`. It does not feel right that the behavior of the zoom depends on the zoom level.

File: examples/panzoom_camera.py [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.drag = None
+
+    def wheelEvent(self, event):  # noqa: N802
+        dim = self.width(), self.height()
+        pos = event.x(), event.y()
+        delta = event.angleDelta().y() * 0.00125
+        view = camera.right - camera.left, camera.top - camera.bottom
+        controls.zoom_to_point(delta, pos, dim, view)
```
Comment (almarklein) [inline examples/panzoom_camera.py:25]: If we provide the control a ref to the camera (instead of only its initial position), then we e.g. would not have to provide the view here ...

File: pygfx/controls/_orbit.py [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+        self.rotation.set_from_rotation_matrix(
+            self._m.look_at(self._v, self._origin, self.up)
+        )
+        return self
+
+    def zoom(self, delta: float) -> "OrbitControls":
+        self.distance -= delta
+        if self.distance < 0:
+            self.distance = 0
+        return self
```
Comment (almarklein) [inline pygfx/controls/_orbit.py:63]: This does not work for orho camera's ;)

Comment (Korijn) [reply pygfx/controls/_orbit.py:63]: it was not intended to be used for ortho cams :p

File: pygfx/cameras/_orthographic.py [hunk 19]
```diff
@@ -33,29 +31,23 @@ def __repr__(self) -> str:
...
+        self.width = float(width)
+        self.height = float(height)
         width = self.width / self.zoom
         height = self.height / self.zoom
-        # Increase eihter the width or height, depending on the view size
-        aspect = width / height
-        if aspect < self._view_aspect:
-            width *= self._view_aspect / aspect
-        else:
-            height *= aspect / self._view_aspect
```
Comment (almarklein) [inline pygfx/cameras/_orthographic.py:47]: Why is this removed? The ortho camera is broken atm, because the content does not scale as the window is resized. Not sure if the missing of these lines are the only problem.

File: pygfx/controls/_orbit.py (hunk 18 above)
Comment (almarklein) [reply pygfx/controls/_orbit.py:63]: But ortho cameras are fine for 3D too, so IMO the orbit controls should work with them.

File: pygfx/cameras/_orthographic.py (hunk 19 above)
Comment (Korijn) [reply pygfx/cameras/_orthographic.py:47]: Haha. Well. Remember this discussion? https://github.com/pygfx/pygfx/pull/53#pullrequestreview-457309496

File: pygfx/controls/_orbit.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (almarklein) [reply pygfx/controls/_orbit.py]: I changed this: the parameter for zoom functions is always a multiplicative value (i.e. a value of 1 does nothing). This makes the zooming feel consistent at every scale.
